Gameplay permissions

dktapps opened the pull request:
## Introduction
This PR implements #5512.

The motivation for this PR is described in #5512.

### Relevant issues
Closes #5512 

## Changes
### API changes
<!-- Any additions to the API that should be documented in release notes? -->

### Behavioural changes
- The following things are now controllable by permissions:
  - Chatting
  - Emoting
  - Placing, breaking and interacting with blocks
  - Dropping, picking up, and activating items
  - Creating new items (e.g. from the creative inventory)
  - Attacking and interacting with entities
  - Flight
  - `CanPlaceOn` and `CanDestroy` checks
  - Creative-style instabreak & destruction of indestructible blocks
- Game modes now apply abilities by setting a permission group to `true` on the player.
- Various places that previously used game mode checks (e.g. dropping items, interacting with blocks, attacking entities) now check for permissions instead.

## Backwards compatibility
Should be fully backwards compatible.

## Follow-up
Add further permissions for the following:
- Opening containers
- Remove some deprecated setters that have been migrated to permissions

## Tests
This has been tested in-game.

## Full list of permission changes per generated documentation

<details>

```diff
diff --git a/source/core-permissions.rst b/../minor-next/core-permissions.rst
index 019e7d9..dc3a2fe 100644
--- a/source/core-permissions.rst
+++ b/../minor-next/core-permissions.rst
@@ -3,7 +3,7 @@
 List of PocketMine-MP core permissions
 ======================================
 
-Generated from PocketMine-MP 5.35.0
+Generated from PocketMine-MP 5.35.1+dev
 
 .. list-table::
    :header-rows: 1
@@ -198,9 +198,81 @@ Generated from PocketMine-MP 5.35.0
    * - ``pocketmine.command.xp.self``
      - Allows the user to add or remove their experience
      - N/A
+   * - ``pocketmine.game.attack.entity``
+     - Allows the user to attack entities
+     - N/A
+   * - ``pocketmine.game.attack.player``
+     - Allows the user to attack other players
+     - N/A
+   * - ``pocketmine.game.block.instabreak``
+     - Allows the user to delete any block without delay, including indestructible blocks
+     - N/A
+   * - ``pocketmine.game.block.mine``
+     - Allows the user to mine blocks
+     - N/A
+   * - ``pocketmine.game.block.place``
+     - Allows the user to place blocks
+     - N/A
+   * - ``pocketmine.game.chat``
+     - Allows the user to chat
+     - N/A
+   * - ``pocketmine.game.emote``
+     - Allows the user to emote
+     - N/A
+   * - ``pocketmine.game.hidden``
+     - Makes the user unseen and unheard by other players (as in spectator game mode)
+     - N/A
+   * - ``pocketmine.game.inventory.creative``
+     - Allows the user to use the creative inventory
+     - N/A
+   * - ``pocketmine.game.inventory.drop``
+     - Allows the user to drop items
+     - N/A
+   * - ``pocketmine.game.inventory.pickup``
+     - Allows the user to pick up items
+     - N/A
+   * - ``pocketmine.game.invulnerable``
+     - Allows the user to ignore all incoming damage (except suicide)
+     - N/A
+   * - ``pocketmine.game.item.bypass.candestroy``
+     - Allows the user to bypass CanDestroy item restrictions when mining blocks
+     - N/A
+   * - ``pocketmine.game.item.bypass.canplaceon``
+     - Allows the user to bypass CanPlaceOn item restrictions when placing blocks
+     - N/A
+   * - ``pocketmine.game.move.flight``
+     - Allows the user to toggle flight mode
+     - N/A
+   * - ``pocketmine.game.move.noclip.block``
+     - Allows the user to pass through blocks
+     - N/A
+   * - ``pocketmine.game.move.noclip.entity``
+     - Allows the user to pass through entities without collision checks
+     - N/A
+   * - ``pocketmine.game.use.block``
+     - Allows the user to use functional blocks like doors
+     - N/A
+   * - ``pocketmine.game.use.entity``
+     - Allows the user to use items on entities
+     - N/A
+   * - ``pocketmine.game.use.item``
+     - Allows the user to use functional items like projectiles
+     - N/A
    * - ``pocketmine.group.console``
      - Grants all console permissions
      - :ref:`Jump<permissions_implied_by_pocketmine.group.console>`
+   * - ``pocketmine.group.gamemode.adventure``
+     - Grants all permissions for actions allowed in the Adventure game mode
+     - :ref:`Jump<permissions_implied_by_pocketmine.group.gamemode.adventure>`
+   * - ``pocketmine.group.gamemode.creative``
+     - Grants all permissions for actions allowed in the Creative game mode
+     - :ref:`Jump<permissions_implied_by_pocketmine.group.gamemode.creative>`
+   * - ``pocketmine.group.gamemode.spectator``
+     - Grants all permissions for actions allowed in the Spectator game mode
+     - :ref:`Jump<permissions_implied_by_pocketmine.group.gamemode.spectator>`
+   * - ``pocketmine.group.gamemode.survival``
+     - Grants all permissions for actions allowed in the Survival game mode
+     - :ref:`Jump<permissions_implied_by_pocketmine.group.gamemode.survival>`
    * - ``pocketmine.group.operator``
      - Grants all operator permissions
      - :ref:`Jump<permissions_implied_by_pocketmine.group.operator>`
@@ -239,6 +311,136 @@ Users granted this permission will also be granted/denied the following permissi
    * - ``pocketmine.group.operator``
      - Granted
 
+.. _permissions_implied_by_pocketmine.group.gamemode.adventure:
+
+Permissions implied by ``pocketmine.group.gamemode.adventure``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Users granted this permission will also be granted/denied the following permissions implicitly:
+
+.. list-table::
+   :header-rows: 1
+
+   * - Name
+     - Type
+   * - ``pocketmine.game.attack.entity``
+     - Granted
+   * - ``pocketmine.game.attack.player``
+     - Granted
+   * - ``pocketmine.game.block.mine``
+     - Granted
+   * - ``pocketmine.game.block.place``
+     - Granted
+   * - ``pocketmine.game.inventory.drop``
+     - Granted
+   * - ``pocketmine.game.inventory.pickup``
+     - Granted
+   * - ``pocketmine.game.use.block``
+     - Granted
+   * - ``pocketmine.game.use.entity``
+     - Granted
+   * - ``pocketmine.game.use.item``
+     - Granted
+
+.. _permissions_implied_by_pocketmine.group.gamemode.creative:
+
+Permissions implied by ``pocketmine.group.gamemode.creative``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Users granted this permission will also be granted/denied the following permissions implicitly:
+
+.. list-table::
+   :header-rows: 1
+
+   * - Name
+     - Type
+   * - ``pocketmine.game.attack.entity``
+     - Granted
+   * - ``pocketmine.game.attack.player``
+     - Granted
+   * - ``pocketmine.game.block.instabreak``
+     - Granted
+   * - ``pocketmine.game.block.mine``
+     - Granted
+   * - ``pocketmine.game.block.place``
+     - Granted
+   * - ``pocketmine.game.inventory.creative``
+     - Granted
+   * - ``pocketmine.game.inventory.drop``
+     - Granted
+   * - ``pocketmine.game.inventory.pickup``
+     - Granted
+   * - ``pocketmine.game.invulnerable``
+     - Granted
+   * - ``pocketmine.game.item.bypass.candestroy``
+     - Granted
+   * - ``pocketmine.game.item.bypass.canplaceon``
+     - Granted
+   * - ``pocketmine.game.move.flight``
+     - Granted
+   * - ``pocketmine.game.use.block``
+     - Granted
+   * - ``pocketmine.game.use.entity``
+     - Granted
+   * - ``pocketmine.game.use.item``
+     - Granted
+
+.. _permissions_implied_by_pocketmine.group.gamemode.spectator:
+
+Permissions implied by ``pocketmine.group.gamemode.spectator``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Users granted this permission will also be granted/denied the following permissions implicitly:
+
+.. list-table::
+   :header-rows: 1
+
+   * - Name
+     - Type
+   * - ``pocketmine.game.hidden``
+     - Granted
+   * - ``pocketmine.game.invulnerable``
+     - Granted
+   * - ``pocketmine.game.move.noclip.block``
+     - Granted
+   * - ``pocketmine.game.move.noclip.entity``
+     - Granted
+
+.. _permissions_implied_by_pocketmine.group.gamemode.survival:
+
+Permissions implied by ``pocketmine.group.gamemode.survival``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Users granted this permission will also be granted/denied the following permissions implicitly:
+
+.. list-table::
+   :header-rows: 1
+
+   * - Name
+     - Type
+   * - ``pocketmine.game.attack.entity``
+     - Granted
+   * - ``pocketmine.game.attack.player``
+     - Granted
+   * - ``pocketmine.game.block.mine``
+     - Granted
+   * - ``pocketmine.game.block.place``
+     - Granted
+   * - ``pocketmine.game.inventory.drop``
+     - Granted
+   * - ``pocketmine.game.inventory.pickup``
+     - Granted
+   * - ``pocketmine.game.item.bypass.candestroy``
+     - Granted
+   * - ``pocketmine.game.item.bypass.canplaceon``
+     - Granted
+   * - ``pocketmine.game.use.block``
+     - Granted
+   * - ``pocketmine.game.use.entity``
+     - Granted
+   * - ``pocketmine.game.use.item``
+     - Granted
+
 .. _permissions_implied_by_pocketmine.group.operator:
 
 Permissions implied by ``pocketmine.group.operator``
@@ -388,4 +590,8 @@ Users granted this permission will also be granted/denied the following permissi
      - Granted
    * - ``pocketmine.command.version``
      - Granted
+   * - ``pocketmine.game.chat``
+     - Granted
+   * - ``pocketmine.game.emote``
+     - Granted
 
```

</details>

dktapps committed:
First look at #5512: gameplay permissions

dktapps committed:
fix CS

dktapps committed:
Merge branch 'minor-next' into gameplay-permissions

dktapps committed:
piece of shit web editor

dktapps committed:
Update DefaultPermissions.php

dktapps committed:
github web editor try not to suck challenge IMPOSSIBLE

dktapps commented:
Blocking items:

- New permissions need descriptions localizing
- Behavioural BC break of `setAllowFlight()` - this can be addressed by removing the base permission during gamemode change - since we'll probably remove `setAllowFlight()` in `major-next`, this workaround is probably OK

dktapps committed:
Merge branch 'minor-next' into gameplay-permissions

dktapps committed:
Fixed behavioural BC break in setAllowFlight()

dktapps committed:
Use permissions for more stuff

dktapps committed:
Added instabreak permission

dktapps committed:
Convert adventure mode checks to permissions

dktapps committed:
Move some gamemode checks to instabreak checks

dktapps committed:
Added noclip permission

dktapps committed:
Remove TODO comment
the remaining bool flags are user input toggles, not abilities, so it doesn't make sense to move them to permissions.

permissions are intended for what the player is *allowed* to do, not what they *want* to do.

dktapps committed:
Added permission for emoting

dktapps committed:
Fix CS

dktapps committed:
Introduce invulnerable permission
this eliminates all remaining usages of isCreative(), except for reach distance checks. Reach distance can't be migrated to a permission in an obvious way.

dktapps committed:
Merge branch 'minor-next' into gameplay-permissions

dktapps committed:
Merge branch 'minor-next' into gameplay-permissions

dktapps committed:
Merge branch 'minor-next' into gameplay-permissions

dktapps commented:
Sticking points I have with this:

- "entity interact" feels non-descriptive and could mean a whole bunch of different things
- Not sure "block delete" is obvious enough (it's for instabreak & breaking bedrock)
- Not sure why we need a permission specifically for right-clicking players

dktapps committed:
Merge branch 'minor-next' into gameplay-permissions

dktapps committed:
Restructure permissions, localise descriptions

dktapps committed:
Add extra permission checks and fix existing ones

kostamax27 commented:
*This probably doesn’t apply to this PR, maybe it’s worth creating **issue** for it?*

It would also be useful to allow granting a player permission for a specific `GameMode` via the `/gamemode` command.

dktapps commented:
There's already an open issue for that. It's not super clear how that would be implemented at the moment.

dktapps committed:
Deprecate Player->isSurvival(), isCreative() et al

these functions have been booby traps for a while since they don't actually do a strict game mode check by default.
In any case, most if not all uses of game mode checks can be superseded by permission checks now.

I think the only cases left uncovered are noclip-entities and spectator mode hiding (not spawned and no sounds sent).

dktapps committed:
Remove final remaining usages of game mode checks

dktapps committed:
Merge branch 'minor-next' into gameplay-permissions

dktapps commented:
My only remaining issue with this is that I'm unsure about the use of permissions for "invulnerable" and "hidden".
Permissions usually operate as gates, but in these cases they're not allowing an action, but just being used as general flags.

Adding permissions for them does enable some nice functionality, but I'm not sure it's the right way to deal with it. I may end up removing those two permissions and sticking with regular flags for the time being.

dktapps committed:
Merge branch 'minor-next' into gameplay-permissions

dktapps committed:
Remove hidden and invulnerable permissions

IvanCraft623 commented:
I agree that invulnerable” and ‘hidden’ should not be treated as permissions.
